ydb/core/kqp/runtime/scheduler/old/kqp_compute_scheduler.cpp

```diff
@@ -187,14 +187,17 @@ class TObservableUpdater {
 
     void CollectValues() {
         std::vector<TParameterKey> toerase;
-        for (auto& [k, v] : Params) {
-            if (!v.Holder->HasDependents()) {
-                toerase.push_back(k);
+        do {
+            toerase.clear();
+            for (auto& [k, v] : Params) {
+                if (!v.Holder->HasDependents()) {
+                    toerase.push_back(k);
+                }
             }
-        }
-        for (auto& key : toerase) {
-            Params.erase(key);
-        }
+            for (auto& key : toerase) {
+                Params.erase(key);
+            }
+        } while (!toerase.empty());
```

Comment:
Looks $O(n^2)$ (unimportant if $n$ is expected to be small and/or this call to be rare)
(besides, it is possible to avoid two passe/`toerase`, just explicitly iterate over map and `it = u.erase(it)`)

Reply:
Тут не так просто избежать квадрата, так как между элементами мапы есть зависимости, по хорошему надо сделать топ сорт по зависимостям и в тком порядке удалять, тогда будет линия

Reply:
Но здесь мало нод - по числу resource пулов (у Cloud Logging он например один)